src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -1405,6 +1428,121 @@ private void WritePackageDependencies()
+
+                        string resolvedPackagePath = _packageResolver.GetPackageDirectory(package.Name, package.Version);
+                        var resolvedPath = resolvedPackagePath ?? string.Empty;
+                        var resolved = !string.IsNullOrEmpty(resolvedPath);
+                        WriteMetadata(MetadataKeys.Resolved, resolved.ToString());
+
+                        string itemPath = package.Path ?? string.Empty;
+                        var path = (resolved
+                            ? resolvedPath
+                            : itemPath) ?? string.Empty;
+                        WriteMetadata(MetadataKeys.Path, path);
+
+                        string itemDiagnosticLevel = GetPackageDiagnosticLevel(package);
+                        var diagnosticLevel = itemDiagnosticLevel ?? string.Empty;
+                        WriteMetadata(MetadataKeys.DiagnosticLevel, diagnosticLevel);
+                    }
+                }
+
+                List<string> GetPackageDependencies()
+                {
+                    List<string> dependencies = new();
+                    HashSet<string> projectFileDependencies = _lockFile.GetProjectFileDependencySet(_compileTimeTarget.Name);
+
+                    var resolvedPackageVersions = _compileTimeTarget.Libraries
+                        .ToDictionary(pkg => pkg.Name, pkg => pkg.Version.ToNormalizedString(), StringComparer.OrdinalIgnoreCase);
+
+                    string frameworkAlias = _lockFile.GetLockFileTargetAlias(_compileTimeTarget);
+
+                    foreach (var package in _compileTimeTarget.Libraries)
+                    {
+                        if (projectFileDependencies.Contains(package.Name))
+                        {
+                            string packageId = GetPackageId(package);
+
+                            // itemSpec
+                            dependencies.Add(packageId);
+                        }
+                    }
+
+                    return dependencies;
+                }
+
+                static string GetPackageId(LockFileTargetLibrary package) => $"{package.Name}/{package.Version.ToNormalizedString()}";
+
+                string GetPackageDiagnosticLevel(LockFileLibrary package)
+                {
+                    string target = _task.TargetFramework ?? "";
+
+                    var messages = _lockFile.LogMessages.Where(log => log.LibraryId == package.Name && log.TargetGraphs
+                                    .Select(tg =>
+                                    {
+                                        var parsedTargetGraph = NuGetFramework.Parse(tg);
+                                        var alias = _lockFile.PackageSpec.TargetFrameworks.FirstOrDefault(tf => tf.FrameworkName == parsedTargetGraph)?.TargetAlias;
+                                        return alias ?? tg;
+                                    }).Contains(target));
+
+                    if (messages.Count() == 0)
```

Comment:
This code double-enumerates `messages`. In the case that there are many messages, you don't need to enumerate all of them here:

```suggestion
                    if (!messages.Any())
```